### PR TITLE
added gocrane support and ecr

### LINKED ISSUE #320: [META] Publish OpenSearch and OpenSearch Dashboards images to Amazon public ECR gallery
**Is your feature request related to a problem? Please describe.**
Currently the OpenSearch and OpenSearch Dashboards bundle images are published only on DockerHub. Customers face throttling issue sometimes on DockerHub.

**Describe the solution you'd like**
Make the below Docker images available on public [ECR repos](https://gallery.ecr.aws/)  under the Amazon verified account so that customers can avoid throttling from Docker Hub

- [ ] [Add the relevant documentation](https://github.com/opensearch-project/documentation-website/issues/631)
- [x] Add support to push to ECR in JenkinsFile
- [x] Copy existing released images from dockerhub staging to ecr staging
- [x] Copy existing released images from dockerhub prod to ecr prod
- [x] https://github.com/opensearch-project/opensearch-build/issues/1518
- [x] https://github.com/opensearch-project/opensearch-ci/issues/65
- [x] #1706
- [x] Make sure ECR images are published from a verified account

**Acceptance Criteria**
- [x] Docker build job should push the docker image on ECR
- [x] Jenkins job to promote the image from staging to production
- [x] ECR repository and roles are deployed using cdk
- [x] Copy existing released images from dockerhub staging to ecr staging
- [x] Copy existing released images from dockerhub prod to ecr prod
- [ ] The ci builds uses the new jenkins job to build docker images

### LINKED ISSUE #1178: Add support to push to ECR in JenkinsFile

### PR DESCRIPTION
**Pull request by abhinavGupta16**
Signed-off-by: Abhinav Gupta <abhng@amazon.com>

### Description
Add support to publish multiarch images to ECR. Change the upload process to dockehub production.

### Current Process
We currently build the image using `docker buildx` and push the image to dockerhub. To push the image to production, the `IS_STAGING` flag is turned off which changes the credentials to production. The image is then rebuilt and uploaded to docker_hub production account.

Since the image is rebuilt, the digest of the image is different, making it impossible to verify if the image in staging is same as production. 

### Desired Solution
Every docker image should only be **built once**. The same image should then be published to all the other accounts - ECR-staging, ECR-production, docker-hub-staging and dockerhub-production.

This would ensure more continuity and reliability in our process and release. It would also improve confidence in the release since the same image will be used for testing and on prodcution.

### Proposed Solution
We can use the [gcrane tool](https://cloud.google.com/artifact-registry/docs/docker/copy-from-gcr) to efficiently copy only the layers required from one account to another. This would ensure that the same image is being promoted from staging to production jenkins, which can be easily verified by the digest hash.

This would help the process in the following ways - 
- The Image being pushed to production would be the same image used in staging. This can be ensured with certainity.
- Image will not have to be rebuilt saving time
- Layers that already exist will be skipped making the copy more efficient

### Pushing the image to ECR
The built image from docker hub should copied over to ECR. Therefore, the digest would remain the same.

### Issues Resolved
#320
#1318 

### Check List
- [X] Commits are signed per the DCO using --signoff 
- [x] R&D roles and policy to upload on ECR
- [x] update existing docker image to include aws-cli for ECR 
- [x] ECR Assume role using jenkins file from within a docker
- [x] upload multiarch image to ECR using jenkinsfile
- [x] R&D possible solutions to promote multiarch images to different docker artifactory and avoid rebuild
- [x] R&D existing docker-build process 
- [x] Add policy and role to cdk deployment
> - opensearch-project/opensearch-ci#52
- [x] Parameters for ECR source and destination image needs to be decided
- [x] ~~Dynamic parameters for staging and prod~~
- [x] ~~Change existing code to accommodate change in `docker-build` job~~
 
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: jenkins/docker/Jenkinsfile
Comment (peternied) [inline jenkins/docker/Jenkinsfile:67]: Please use the built-in functions for downloading files.  Minimal shell script should always be the goal.

File: jenkins/docker/Jenkinsfile
Comment (peternied) [inline jenkins/docker/Jenkinsfile:32]: Looks like these parameters are derived from another parameter, lets move this logic into code rather than force the caller to know which field to populate

File: jenkins/docker/Jenkinsfile
Comment (peternied) [inline jenkins/docker/Jenkinsfile:82]: While we are investing in this code, it is time to fix `$DOCKER_BUILD_SCRIPT_WITH_COMMANDS`.  Lets break this apart into the input components:
- OpenSearch x64/arm64 artifacts
- destination tag
- docker repository target

File: jenkins/docker/Jenkinsfile
Comment (peternied) [inline jenkins/docker/Jenkinsfile:81]: We should never manually log details about credentials object being used.  I believe Jenkins has logging that includes properly masked credentials info, if we need additional logging can you see what the gaps are so we can fix the plugins or wrap them properly.

File: jenkins/docker/Jenkinsfile
Comment (peternied) [inline jenkins/docker/Jenkinsfile:108]: Nice role and role name :D

File: jenkins/docker/Jenkinsfile
Comment (peternied) [inline jenkins/docker/Jenkinsfile:110]: Duplicate code.  Lets create a groovy library for this.

File: jenkins/docker/Jenkinsfile
Comment (peternied) [inline jenkins/docker/Jenkinsfile:118]: Seems like these should be parameterized, why the double copy?

**Review by peternied — COMMENTED** [review]
Great start - how long does the additional process take?  

Were you able to update the deployment cdk for this, or do we need to make changes in opensearch-ci?

File: jenkins/docker-ecr/Jenkinsfile
Comment (peterzhuamazon) [inline jenkins/docker-ecr/Jenkinsfile:151]: This should be directly as part of the ci-runner image.

Comment (peternied) [reply jenkins/docker-ecr/Jenkinsfile:151]: Instead of continuing to add onto a 'central' image we should create a separate images for each type of task.

File: jenkins/docker-ecr/Jenkinsfile
Comment (peternied) [inline jenkins/docker-ecr/Jenkinsfile:63]: We should never run jobs as root, instead we should correct permissions issues on the image so it can be used without this go around.

File: jenkins/docker-ecr/Jenkinsfile
Comment (peternied) [inline jenkins/docker-ecr/Jenkinsfile:114]: In ECR what benefits do we have for using more than one account?  `production` vs `staging` in our docker images is not much of a distinction as from an infrastructure perspective both are production systems for considerations of data loss or up-time.

**Review by peternied — CHANGES_REQUESTED** [review]
We can change this job to accept a docker image path and a destination path, so it only copies the image over to docker/ECR.  That would let us boil down this file to the plumbing needed to do `gcrane cp $DOCKER_SRC $DOCKER_DEST|$ECR_DEST`.

Further thoughts; How can we integrate this job as part of the promotion job?

**Comment by abhinavGupta16** [comment]
> Were you able to update the deployment cdk for this, or do we need to make changes in opensearch-ci?

Docker does not work in `opensearch-ci`, so no job would work there currently. Hence, it does need tinkering.

**Comment by abhinavGupta16** [comment]
> Great start - how long does the additional process take?

I did not time it, but not more than a few minutes. So it's good!

File: jenkins/docker-ecr/Jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker-ecr/Jenkinsfile:151]: I agree with @peternied since this is very specific to this task only. But would we not end up creating a lot of images over a period of time? We need to come up with some plan to follow. Thoughts?

File: jenkins/docker-ecr/Jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker-ecr/Jenkinsfile:63]: Ideally, yes. However, this is how it's done in the current `docker-build` job. Maybe @peterzhuamazon can shed light as to why?

File: jenkins/docker-ecr/Jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker-ecr/Jenkinsfile:114]: The docker images we release currently are released as staging and production. We should have this distinction in ECR as well.

We should not have data loss anywhere, production and staging. However, I believe production failure is much more critical than staging.

File: jenkins/docker/Jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker/Jenkinsfile]: was just testing out a few things there

File: jenkins/docker/Jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker/Jenkinsfile]: created a function for it.

File: jenkins/docker/Jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker/Jenkinsfile]: lol. That's just for testing. I'll update it when it's ready for review.

File: jenkins/docker/Jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker/Jenkinsfile]: This was a part of `docker-build` job. But I can remove it from this

File: jenkins/docker/Jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker/Jenkinsfile]: You have a point. However, would that not make the job specific to opensearch only?

File: jenkins/docker/Jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker/Jenkinsfile]: Updated the parameters

**Comment by abhinavGupta16** [comment]
https://github.com/opensearch-project/opensearch-ci/issues/52 - Issue to create role and policy for ECR via cdk

File: jenkins/docker-ecr/Jenkinsfile
Comment (peternied) [reply jenkins/docker-ecr/Jenkinsfile:143]: I think that we shouldn't update the a central CI image.

However, running curl/install each execution is can running this job with an additional dependency that can fail.  I am not worried about the number of images so long as how they are build has the docker file and a way to search them.  

https://github.com/opensearch-project/opensearch-build/issues/1207 would be a good place for more discussion

File: jenkins/docker-ecr/Jenkinsfile
Comment (peternied) [reply jenkins/docker-ecr/Jenkinsfile]: Note; We discussed this offline, the names of these environments staging and production are not uniformly applied which creates some confusing.  Rather than have boolean's like 'is_production' lets accept paths and credential references to allow  for these to be placed in a deterministic way

**Comment by abhinavGupta16** [comment]
Added issue https://github.com/opensearch-project/opensearch-build/issues/1516 to copy image from staging to production

File: jenkins/docker-ecr/Jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker-ecr/Jenkinsfile]: Following up on our discussion today, I have created an issue to support production promotion of artifacts via the promotion job

File: jenkins/docker-ecr/docker-build-ecr.jenkinsfile
Comment (peternied) [inline jenkins/docker-ecr/docker-build-ecr.jenkinsfile:48]: Since this value is reused, lets put it in a variable so it isn't repeated

**Review by peternied — COMMENTED** [review]
Much more condensed and flexible, thanks for all those updates.  Small number of follow up items from previous revisions

File: jenkins/docker-ecr/docker-build-ecr.jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker-ecr/docker-build-ecr.jenkinsfile]: @peterzhuamazon

File: jenkins/docker-ecr/docker-build-ecr.jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker-ecr/docker-build-ecr.jenkinsfile:11]: We need root for buildx and installing gcrane

File: jenkins/docker-ecr/docker-build-ecr.jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker-ecr/docker-build-ecr.jenkinsfile]: The command helps run the build script for docker from any repository. Changed the params and jobs now

**Comment by abhinavGupta16** [comment]
Created an issue to integrate this job with `buildDockerImage.groovy` https://github.com/opensearch-project/opensearch-build/issues/1518

**Review by peternied — APPROVED** [review]
Lets not hold up this PR on those outstanding items, please create issues for any you can't resolve at this time so we can move forward with these changes.

**Comment by abhinavGupta16** [comment]
> Lets not hold up this PR on those outstanding items, please create issues for any you can't resolve at this time so we can move forward with these changes.

I created the issues mentioned in prev comments - #1518, #1516, opensearch-project/opensearch-ci#52

File: jenkins/docker-ecr/docker-build-ecr.jenkinsfile
Comment (peterzhuamazon) [reply jenkins/docker-ecr/docker-build-ecr.jenkinsfile]: It will.

Comment (peterzhuamazon) [reply jenkins/docker-ecr/docker-build-ecr.jenkinsfile]: if it is invoking the agent node socket, in which only docker builder image is doing right now.

File: jenkins/docker-ecr/docker-promote.jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker-ecr/docker-promote.jenkinsfile:28]: How about renaming it to `dockerhub-promote-to-prod`? Promote to prod would implicitly mean copying over from staging to prod. Copy asks copy from where to where

Comment (peterzhuamazon) [reply jenkins/docker-ecr/docker-promote.jenkinsfile:28]: I am ok with promote I just dont think build should be part of it.

File: jenkins/docker-ecr/docker-build-ecr.jenkinsfile
Comment (gaiksaya) [inline jenkins/docker-ecr/docker-build-ecr.jenkinsfile:26]: Are we expecting to build docker images from repositories other than opensearch-build? If not today we can add this later when it is required
trying to see if we can reduce the parameters.

File: jenkins/docker-ecr/docker-build-ecr.jenkinsfile
Comment (gaiksaya) [inline jenkins/docker-ecr/docker-build-ecr.jenkinsfile:28]: I believe the default should be something that if the value is not provided by the user it should run with default value. This is more of an example that should go in the description. Please remove default if you don't have any

File: jenkins/docker-ecr/docker-promote.jenkinsfile
Comment (gaiksaya) [inline jenkins/docker-ecr/docker-promote.jenkinsfile:22]: typo: staging

File: vars/copyDockerImage.groovy
Comment (gaiksaya) [inline vars/copyDockerImage.groovy:15]: Is it always going to be `us-east-1` or should we take this as a param? Your call

Comment (abhinavGupta16) [reply vars/copyDockerImage.groovy:15]: I have the repositories created on `us-east-1` as of today

File: jenkins/docker-ecr/docker-build-ecr.jenkinsfile
Comment (abhinavGupta16) [reply jenkins/docker-ecr/docker-build-ecr.jenkinsfile:26]: I had a conversation with @peterzhuamazon about it. We need the repo etc to specify which repo has the `build.sh` script. For example, this is also used for `data-prepper`, and they have their build script in their repo

Comment (gaiksaya) [reply jenkins/docker-ecr/docker-build-ecr.jenkinsfile:26]: I do not see `build.sh` script anywhere in this file. It is not generic enough today

Comment (abhinavGupta16) [reply jenkins/docker-ecr/docker-build-ecr.jenkinsfile:26]: The`build.sh` would be pulled in from the repo that is listed here. Eg: 

https://github.com/opensearch-project/opensearch-build/blob/main/docker/release/build-image-multi-arch.sh

https://github.com/opensearch-project/data-prepper/blob/main/release/docker/ci/build-image-single-arch.sh

The command would then specify how to run it and parameters for the script

**Review by peterzhuamazon — APPROVED** [review]
I generally dont have an issue with this PR as a start.
I do want @gaiksaya to also give an approval before we approve this.
Since this is quite a big one added to our system.

File: jenkins/docker-ecr/docker-build-ecr.jenkinsfile
Comment (gaiksaya) [reply jenkins/docker-ecr/docker-build-ecr.jenkinsfile]: Ahh! Its the naming convention issue then. Its a nit maybe fix it later. Moving on

**Review by gaiksaya — APPROVED** [review]

**Pull request merged**
